src/esp/gfx/CubeMap.cpp

```diff
@@ -0,0 +1,399 @@
+  Cr::Containers::Pointer<Mn::Trade::AbstractImageConverter> converter;
+  if (!(converter = manager.loadAndInstantiate("AnyImageConverter"))) {
+    return false;
+  }
+
+  std::string coordStrings[6] = {".+X", ".-X", ".+Y", ".-Y", ".+Z", ".-Z"};
+  for (int iFace = 0; iFace < 6; ++iFace) {
+    // TODO: use framebuffer.read() instead somehow?
+    // But the question is it has NOTHING to do with the framebuffer
+    // anymore at this moment.
+    Mn::Image2D image = textures_[type]->image(
+        convertFaceIndexToCubeMapCoordinate(iFace), 0, {getPixelFormat(type)});
+
+    std::string filename;
+    switch (type) {
+      case TextureType::Color: {
+        filename = imageFilePrefix + getTextureTypeFilenameString(type) +
+                   coordStrings[iFace] + std::string{".png"};
+      } break;
+
+      case TextureType::Depth: {
+        filename = imageFilePrefix + getTextureTypeFilenameString(type) +
+                   coordStrings[iFace] + std::string{".hdr"};
+      } break;
+
+      default:
+        CORRADE_INTERNAL_ASSERT_UNREACHABLE();
+        break;
+    }
+
+    if (!converter->exportToFile(image, filename)) {
+      return false;
+    } else {
+      LOG(INFO) << "Saved image " << iFace << " to " << filename;
+    }
+  }
+
+  return true;
+#endif
+}
+
+void CubeMap::renderToTexture(CubeMapCamera& camera,
+                              scene::SceneGraph& sceneGraph,
+                              RenderCamera::Flags flags) {
+  CORRADE_ASSERT(camera.isInSceneGraph(sceneGraph),
+                 "CubeMap::renderToTexture(): camera is NOT attached to the "
+                 "current scene graph.", );
+  // ==== projection matrix ====
+  // CAREFUL! In this function the projection matrix of the camera is assumed to
+  // be set already outside of this function by the user.
+  // we simply do sanity check here.
+  {
+    Mn::Vector2i vp = camera.viewport();
+    CORRADE_ASSERT(vp.x() == vp.y() && vp.x() == imageSize_,
+                   "CubeMap::renderToTexture(): the image size with in the "
+                   "CubeMapCamera, which is"
+                       << vp.x() << "by" << vp.y() << "compared to"
+                       << imageSize_ << "is not correct.", );
+  }
+
+  // ==== camera matrix ====
+  // In case user set the relative transformation of
+  // the camera node before calling this function, original viewing matrix of
+  // the camera MUST be updated as well.
+  camera.updateOriginalViewingMatrix();
+  frameBuffer_.bind();
+  for (int iFace = 0; iFace < 6; ++iFace) {
+    camera.switchToFace(iFace);
+    prepareToDraw(iFace);
+
+    // TODO:
+    // camera should have flags so that it can do "low quality" rendering,
+    // e.g., no normal maps, no specular lighting, low-poly meshes,
+    // low-quality textures.
+
+    for (auto& it : sceneGraph.getDrawableGroups()) {
+      // TODO: remove || true
+      if (it.second.prepareForDraw(camera) || true) {
```

Comment:
Are you asking `it.second.prepareForDraw(camera)`?
This is not something newly invented. I just copied it from any drawEvent().

Reply:
Yeah I was wondering what is the reason for `prepareForDraw()` failing, what can cause that? Is it even valid for it to fail here? Shouldn't it be asserted? etc.

Reply:
Just look into this function:
```
  /**
   * @return Whether the @ref DrawableGroup is in a valid state to be drawn
   */
  virtual bool prepareForDraw(const RenderCamera&) { return true; }
```

I guess in the future, we may have different drawable groups, and not all of them would be drawn.

Reply:
Hmm. Doesn't really *feel* like a good API design to me (because I had to ask what was the actual meaning). 

Nevertheless, this means the `|| true` should then be removed, as your TODO says.

Reply:
I guess Matthew and Alex put it here last year. Will need to sync up with them first.